src/scale_impls/tracing_decoder/visitor.rs

```diff
@@ -0,0 +1,365 @@
+// Copyright (C) 2022-2024 Parity Technologies (UK) Ltd. (admin@parity.io)
+// This file is a part of the scale-value crate.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//         http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::prelude::*;
+
+use super::error::TraceDecodingError;
+use super::path::{Path, PathSegment};
+use crate::{Composite, Primitive, Value, ValueDef, Variant};
+use core::marker::PhantomData;
+use scale_decode::visitor::TypeIdFor;
+use scale_type_resolver::TypeResolver;
+
+/// A visitor that will attempt to decode some bytes into a [`crate::Value`],
+/// returning a detailed error of where the decoding fails if it does.
+pub struct TraceDecodingVisitor<Resolver> {
+    path: Path,
+    marker: PhantomData<Resolver>,
+}
+
+impl<Resolver> TraceDecodingVisitor<Resolver> {
+    fn at(&self, segment: PathSegment) -> Self {
+        TraceDecodingVisitor { path: self.path.at(segment), marker: PhantomData }
+    }
+    fn at_idx(&self, idx: usize) -> Self {
+        self.at(PathSegment::Index(idx))
```

Comment:
`self.path.at_index()`?

Reply:
I did do that initially, but then you also have to wrap that path into a new TraceDecodiongVisitor so it ends up being more lines of code!

Reply:
Actually I did something like this in the end, because it is def nicer not to expose `PathSegment` to keep the things exposed in path.rs cleaner :)